blockchain/vm/evm.go

```diff
@@ -511,16 +511,23 @@ func (evm *EVM) create(caller types.ContractRef, codeAndHash *codeAndHash, gas u
 		evm.StateDB.AddAddressToAccessList(address)
 	}
 
-	// Ensure there's no existing contract already at the designated address
+	// Ensure there's no existing contract already at the designated address.
+	// Account is regarded as existent if any of these three conditions is met:
+	// - the nonce is nonzero
+	// - the code is non-empty
+	// - the storage is non-empty
 	contractHash := evm.StateDB.ResolveCodeHash(address)
+	storageRoot := evm.StateDB.GetStorageRoot(address)
 
 	// The early Kaia design tried to support the account creation with a user selected address,
 	// so the account overwriting was restricted.
 	// Because the feature was postponed for a long time and the restriction can be abused to prevent SCA creation,
 	// Kaia enables SCA overwriting over EOA like Ethereum after Shanghai compatible hardfork.
 	// NOTE: The following code should be re-considered when Kaia enables TxTypeAccountCreation
 	if evm.chainRules.IsShanghai {
-		if evm.StateDB.GetNonce(address) != 0 || (contractHash != (common.Hash{}) && contractHash != emptyCodeHash) {
+		if evm.StateDB.GetNonce(address) != 0 ||
+			(contractHash != (common.Hash{}) && contractHash != types.EmptyCodeHash) || // non-empty code
+			(storageRoot != (common.Hash{}) && storageRoot != types.EmptyRootHash) { // non-empty storage
```

Comment:
Assume `addr` is EOA with code, then `GetStorageRoot(addr)` is zero, but its nonce is non-zero so it's blocked, right? Could you mention it in the comment?

Reply:
Why does the `GetStorageRoot(addr)` become zero if `addr` is EOA with code? It will be either an empty or non-empty calculated root.

Anyway, if you mean `empty` root, yes, it will be blocked since nonce isn't zero.

Reply:
Doesn't type assertion in `GetProgramAccount(addr)` fail, making it return nil?

Reply:
I thought the type assertion for `ProgramAccount` becomes all true for both EOA and SCA since the implementation of EIP-7702. Please check this https://github.com/kaiachain/kaia/blob/dev/blockchain/state/statedb_test.go#L229.

Reply:
oh I see. EOA implements ProgramAccount (check by adding below test here)
https://github.com/kaiachain/kaia/blob/0bada75bcb205593ec28c3bd268fa8f5c00b2802/blockchain/types/account/account_test.go#L42.

```
	_ ProgramAccount = (*ExternallyOwnedAccount)(nil)
```